blog/_posts/2021-12-14-sustainable-scala.md

```diff
@@ -0,0 +1,278 @@
+only changed some patterns that we believe are typically non-idiomatic.
+
+For instance, the "optimized" version of the `binary-trees` benchmark models 
+an empty tree as tree whose branches are `null`:
+
+~~~ scala
+final case class Tree(left: Tree, right: Tree) {
+  def checkSum: Int =
+    left match {
+      case null => 1
+      case tl   => 1 + tl.checkSum + right.checkSum
+    }
+}
+
+object Tree {
+  final val EmptyTree = Tree(null, null)
+}
+~~~
+
+Whereas the idiomatic one uses a class hierarchy:
+
+~~~ scala
+sealed trait Tree {
+  def checkSum: Int
+}
+
+case class NonEmptyTree(left: Tree, right: Tree) extends Tree {
+  def checkSum: Int = 1 + left.checkSum + right.checkSum
+}
+
+case object EmptyTree extends Tree {
+  def checkSum: Int = 1
+}
+~~~
+
+Does this change impact performance? Keep reading to know the answer.
+
+Another example of difference between idiomatic and optimized versions is the 
+usage of Scala collections instead of Java `Array`s, and the usage of `for` 
+loop instead of `while` loops.
+
+### Protocol
+
+We compared the energy consumption of 10 benchmarks, each written in C, Scala, 
+Java, JavaScript, and Python. To achieve this, we executed each benchmark 10 
+times (after 5 warm-up iterations), on an Intel i9-7900X @ 3.30 GHz with 20 
+CPUs and 128 GB of memory, with OpenJDK 17, Node 10.19.0, and Python 3.8.10.
+
+You can find the source code of the benchmarks in the following Git repository:
+[https://github.com/WojciechMazur/Energy-Languages](https://github.com/WojciechMazur/Energy-Languages/tree/feature/scala-develop).
+
+### Results
+
+To visualize better how the languages compare (regardless of the nature of 
+the benchmarks), we normalized the measurements, using C as a baseline.
+
+The figure below shows the normalized average energy consumption for each 
+benchmark, for the languages C, Java, and Scala (lower is better):
+
+![](/resources/img/blog/sustainable-scala/c-java-scala-1.png)
+![](/resources/img/blog/sustainable-scala/c-java-scala-2.png)
+
+We observe that the Scala benchmarks were sometimes consuming a similar 
+amount of energy as the C benchmark (this is the case for `binary-trees`, 
+`fannkuch-redux`, `fasta`, and `n-body`), and sometimes a significantly higher 
+amount of energy (up to 12 times more energy for `regex-redux`).
+
+The figure below shows the same information, but it now includes JavaScript 
+and Python:
+
+![](/resources/img/blog/sustainable-scala/c-java-scala-js-python-1.png)
+![](/resources/img/blog/sustainable-scala/c-java-scala-js-python-2.png)
+
+Compared to C, the Python benchmarks consumed between 4 to 339 times more 
+energy, and the JavaScript benchmarks consumed between 2 to 12 times 
+more energy.
+
+Last, the figure below compares C, Java, Scala, and "idiomatic" Scala 
+benchmarks:
+
+![](/resources/img/blog/sustainable-scala/c-java-scala-idiomatic-1.png)
+![](/resources/img/blog/sustainable-scala/c-java-scala-idiomatic-2.png)
```

Comment:
It would be easier to read if `Scala` always has the same color in all 6 graphs.

Reply:
Done in f55c814